File: Backend/Nodejs/src/models/song.ts
import { pool } from "./database/connection";
import { Song, SongFiles, UpdateSong } from "./types";

export class SongModel {
  static createSong(
    song: Song,
    files: SongFiles,
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      const newSong = { ...song };
      newSong.cover = files.cover[0].location;
      newSong.source = files.source[0].location;
      pool.query("CALL CreateSong(?,?,?,?,?)", [
        newSong.name,
        newSong.cover,
        newSong.duration,
        newSong.artist,
        newSong.source,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getSong(
    { id, email }: { id: number; email: string },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query(
        `CALL GetSong(?,?)`,
        [id, email],
        (err, result) => {
          if (err) throw err;
          callback(result[0], true);
        },
      );
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getSongs(callback: (response: any, ok: Boolean) => void) {
    try {
      pool.query(
        `CALL GetAllSongs()`,
        (err, result) => {
          if (err) throw err;
          callback(result[0], true);
        },
      );
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static editSong(
    newSong: UpdateSong,
    files: SongFiles,
    callback: Function,
  ) {
    try {
      newSong.cover = files.cover === undefined ? "" : files.cover[0].location;
      newSong.source = files.source === undefined
        ? ""
        : files.source[0].location;

      // Encriptando
      // Guardar en DB
      pool.query("CALL UpdateSong(?,?,?,?,?,?)", [
        newSong.id,
        newSong.name,
        newSong.cover,
        newSong.duration,
        newSong.artist,
        newSong.source,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static deleteSong(
    { id }: { id: number },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL DeleteSong(?)", [
        id,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }
}

File: Backend/Python/src/models/song.py
from src.models.database.connection import getCnx


class SongModel:

    @staticmethod
    def create_song(song, cover_filename, source_filename):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "CreateSong",
                (
                    song["name"],
                    cover_filename,
                    int(song["duration"]),
                    int(song["artist"]),
                    source_filename,
                ),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    return result, False
                else:
                    return result, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def get_song(id, email):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetSong",
                (
                    id,
                    email,
                ),
            )
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    result = dict(
                        zip(
                            ("id", "name", "singer", "cover", "musicSrc", "inFav"), song
                        )
                    )
                    data.append(result)
            return data, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def get_songs():
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetAllSongs",
            )
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    result = dict(
                        zip(
                            ("id", "name", "cover", "musicSrc", "artist", "album"), song
                        )
                    )
                    data.append(result)
            return data, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def delete_song(id_song):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "DeleteSong",
                (int(id_song),),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    return result, False
                else:
                    return result, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def edit_song(song, cover_filename, source_filename):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "Updatesong",
                (
                    song["id"],
                    song["name"],
                    cover_filename,
                    song["duration"],
                    song["artist"],
                    source_filename,
                ),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    return result, False
                else:
                    return result, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()
